tamr_unify_client/client.py

```diff
@@ -1,62 +1,76 @@
+import logging
+from typing import Optional
 from urllib.parse import urljoin
 
 import requests
-from requests import Response
+import requests.auth
+import requests.exceptions
 
 from tamr_unify_client.dataset.collection import DatasetCollection
 from tamr_unify_client.project.collection import ProjectCollection
 
-# monkey-patch Response.successful
+logger = logging.getLogger(__name__)
 
 
-def successful(self):
-    """Checks that this response did not encounter an HTTP error (i.e. status code indicates success: 2xx, 3xx).
+def successful(response: requests.Response) -> requests.Response:
+    """Checks that this response did not encounter an HTTP error.
 
-    :raises :class:`requests.exceptions.HTTPError`: If an HTTP error is encountered.
-    :return: The calling response (i.e. ``self``).
-    :rtype: :class:`requests.Response`
+    HTTP error codes match 4xx or 5xx.
+
+    Returns:
+        The response being checked.
+
+    Raises:
+        requests.exceptions.HTTPError: If an HTTP error is encountered.
     """
-    self.raise_for_status()
-    return self
+    try:
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as e:
+        logger.error(f"HTTP error code response body: {e.response.text}")
+        raise e
+    return response
+
 
+# monkey-patch requests.Response.successful
+requests.Response.successful = successful
 
-Response.successful = successful
+
+def _default_response_message(response: requests.Response) -> str:
+    req = response.request
+    return f"{req.method} {response.url} : {response.status_code}"
```

Comment:
Does this conform to an existing logging standard?

Reply:
No. I did some digging and found the [format used by `urllib3`](https://github.com/urllib3/urllib3/blob/75d271954b5d662d0567e95b054475e208619dff/src/urllib3/connectionpool.py#L428) (lib used under-the-hood by `requests`).

```python
f'{scheme}://{host}:{port} "{method} {url} HTTP/{http_version}" {status_code} {content_length}'
```

We could crib that....

OR we could just use the logging they already set up instead of writing our own as [shown here](https://2.python-requests.org/en/master/api/#api-changes) (bottom most subsection of "API Changes"). But it looks pretty gnarly...

Reply:
Not only is the setup for `urllib3` logging non-trivial, but you have to set logging level to `DEBUG` to get the request/response info which means you end up getting a lot of cruft:

```sh
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): httpbin.org:443
send: b'GET /headers/jflajsdflkjj HTTP/1.1\r\nHost: httpbin.org\r\nUser-Agent: python-requests/2.22.0\r\nAccept-Encoding: gzip, deflate\r\nAccept: */*\r\nConnection: keep-alive\r\n\r\n'
reply: 'HTTP/1.1 404 NOT FOUND\r\n'
header: Access-Control-Allow-Credentials: true
header: Access-Control-Allow-Origin: *
header: Content-Encoding: gzip
header: Content-Type: text/html
header: Date: Thu, 19 Dec 2019 20:39:31 GMT
header: Referrer-Policy: no-referrer-when-downgrade
header: Server: nginx
header: X-Content-Type-Options: nosniff
header: X-Frame-Options: DENY
header: X-XSS-Protection: 1; mode=block
header: Content-Length: 198
header: Connection: keep-alive
DEBUG:urllib3.connectionpool:https://httpbin.org:443 "GET /headers/jflajsdflkjj HTTP/1.1" 404 198
```